src/gpu_sysman_test.c

```diff
@@ -68,6 +68,25 @@
 #define FP_LAYOUT_NEED_NOTHING 1
 #include "gpu_sysman.c" /* test this */
 
+/* utilities needed from collectd core */
+#include <unistd.h>
+/* common.c fails to -Werror=unused-parameter, so
```

Comment:
Can you post the error you're getting? This sounds like it would be easy to fix.

Reply:
`getpwnam_r()` implementation in `common.c` does not use all of its parameters.  I worked around that by using `#define HAVE_GETPWNAM_R 1`.

After that, and fixing the includes, there are still following warnings:
```
gcc -Werror -Wall -Wextra -Wpedantic -Wcast-align=strict -Wformat-security -Wnull-dereference -Wstrict-overflow=4 -Warray-bounds=2 -Wno-aggressive-loop-optimizations -I. -Idaemon -O3 -g -D_FORTIFY_SOURCE=2 gpu_sysman_test.c -o gpu_sysman_test
In file included from gpu_sysman_test.c:74:
utils/common/common.c: In function ‘check_create_dir’:
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 256 equals destination size [-Werror=stringop-truncation]
   97 |   strncpy(dest, src, n);
      |   ^
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 256 equals destination size [-Werror=stringop-truncation]
utils/common/common.c: In function ‘notification_init’:
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 256 equals destination size [-Werror=stringop-truncation]
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 128 equals destination size [-Werror=stringop-truncation]
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 128 equals destination size [-Werror=stringop-truncation]
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 128 equals destination size [-Werror=stringop-truncation]
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 128 equals destination size [-Werror=stringop-truncation]
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 128 equals destination size [-Werror=stringop-truncation]
utils/common/common.c: In function ‘notification_init_metric’:
utils/common/common.c:97:3: error: ‘__builtin_strncpy’ specified bound 256 equals destination size [-Werror=stringop-truncation]
utils/common/common.c: In function ‘walk_directory’:
utils/common/common.c:1142:5: error: assuming signed overflow does not occur when changing X +- C1 cmp C2 to X cmp C2 -+ C1 [-Werror=strict-overflow]
 1142 | int walk_directory(const char *dir, dirwalk_callback_f callback,
      |     ^~~~~~~~~~~~~~
```

Reply:
Which version of GCC is that? I have been unable to reproduce with GCC on Debian 12:

```
$ make V=1 src/utils/common/common.lo
depbase=`echo src/utils/common/common.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
/bin/bash ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I./src  -I./src -I./src/daemon -DPREFIX='"/tmp/collectd"' -DCONFIGFILE='"/tmp/collectd/etc/collectd.conf"' -DLOCALSTATEDIR='"/tmp/collectd/var"' -DPKGLOCALSTATEDIR='"/tmp/collectd/var/lib/collectd"' -DPLUGINDIR='"/tmp/collectd/lib/collectd"' -DPKGDATADIR='"/tmp/collectd/share/collectd"'   -Wall -Werror -Werror -Wall -Wextra -Wpedantic -Wcast-align=strict -Wformat-security -Wnull-dereference -Wstrict-overflow=4 -Warray-bounds=2 -Wno-aggressive-loop-optimizations -MT src/utils/common/common.lo -MD -MP -MF $depbase.Tpo -c -o src/utils/common/common.lo src/utils/common/common.c &&\
mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I./src -I./src -I./src/daemon -DPREFIX=\"/tmp/collectd\" -DCONFIGFILE=\"/tmp/collectd/etc/collectd.conf\" -DLOCALSTATEDIR=\"/tmp/collectd/var\" -DPKGLOCALSTATEDIR=\"/tmp/collectd/var/lib/collectd\" -DPLUGINDIR=\"/tmp/collectd/lib/collectd\" -DPKGDATADIR=\"/tmp/collectd/share/collectd\" -Wall -Werror -Werror -Wall -Wextra -Wpedantic -Wcast-align=strict -Wformat-security -Wnull-dereference -Wstrict-overflow=4 -Warray-bounds=2 -Wno-aggressive-loop-optimizations -MT src/utils/common/common.lo -MD -MP -MF src/utils/common/.deps/common.Tpo -c src/utils/common/common.c  -fPIC -DPIC -o src/utils/common/.libs/common.o
$ gcc --version
gcc (Debian 12.2.0-14) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Reply:
#4233 attempts to fix this.

Reply:
This was with the  default GCC version in Ubuntu 22.04 LTS:
`gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04)`

However, to get GCC to actually analyze the code more thoroughly, you also need to use high optimization levels (I used `-O3`, your command line does not seem to  specify any optimizations?).

And your command line was missing `-D_FORTIFY_SOURCE=2`: https://wiki.ubuntu.com/Security/Features#Built_with_Fortify_Source